specifications/tcp/tcp.tla

```diff
@@ -0,0 +1,330 @@
+        /\ connstate = [p \in Peers |-> IF p = node THEN "SYN-SENT" ELSE "LISTEN"]
+        /\ network = [p \in Peers |-> IF p = node THEN <<>> ELSE << "SYN" >>]
+
+\* Action triggered by user commands:
+
+PASSIVE_OPEN(local, remote) ==
+    \* User command: PASSIVE-OPEN
+    /\ local # remote
+    /\ connstate[local] = "CLOSED"
+    /\ UNCHANGED network
+    /\ connstate' = [connstate EXCEPT ![local] = "LISTEN"]
+    /\ tcb' = [tcb EXCEPT ![local] = TRUE]
+
+ACTIVE_OPEN(local, remote) ==
+    \* User command: ACTIVE-OPEN
+    /\ local # remote
+    /\ connstate[local] = "CLOSED"
+    /\ network' = [ network EXCEPT ![remote] = Append(@, "SYN")]
+    /\ connstate' = [connstate EXCEPT ![local] = "SYN-SENT"]
+    /\ tcb' = [tcb EXCEPT ![local] = TRUE]
+
+CLOSE_SYN_SENT(local, remote) ==
+    \* User command: CLOSE
+    /\ local # remote
+    /\ connstate[local] = "SYN-SENT"
+    /\ UNCHANGED network
+    /\ connstate' = [connstate EXCEPT ![local] = "CLOSED"]
+    /\ tcb' = [tcb EXCEPT ![local] = FALSE]
+
+CLOSE_SYN_RECEIVED(local, remote) ==
+    \* User Command: CLOSE
+    /\ local # remote
+    /\ connstate[local] = "SYN-RECEIVED"
+    /\ network' = [ network EXCEPT ![remote] = Append(@, "FIN")]
+    /\ connstate' = [connstate EXCEPT ![local] = "FIN-WAIT-1"]
+    /\ UNCHANGED tcb
+
+SEND(local, remote) ==
+    \* User command: SEND
+    /\ local # remote
+    /\ connstate[local] = "LISTEN"
+    /\ network' = [ network EXCEPT ![remote] = Append(@, "SYN")]
+    /\ connstate' = [connstate EXCEPT ![local] = "SYN-SENT"]
+    /\ UNCHANGED tcb
+
+CLOSE_LISTEN(local, remote) ==
+    \* User command: CLOSE
+    /\ local # remote
+    /\ connstate[local] = "LISTEN"
+    /\ UNCHANGED network
+    /\ connstate' = [connstate EXCEPT ![local] = "CLOSED"]
+    /\ tcb' = [tcb EXCEPT ![local] = FALSE]
+
+CLOSE_ESTABLISHED(local, remote) ==
+    \* User command: CLOSE
+    /\ local # remote
+    /\ connstate[local] = "ESTABLISHED"
+    /\ network' = [ network EXCEPT ![remote] = Append(@, "FIN")]
+    /\ connstate' = [connstate EXCEPT ![local] = "FIN-WAIT-1"]
+    /\ UNCHANGED tcb
+
+CLOSE_CLOSE_WAIT(local, remote) ==
+    \* User command: CLOSE
+    /\ local # remote
+    /\ connstate[local] = "CLOSE-WAIT"
+    /\ network' = [ network EXCEPT ![remote] = Append(@, "FIN")]
+    /\ connstate' = [connstate EXCEPT ![local] = "LAST-ACK"]
+    /\ UNCHANGED tcb
+
+User ==
+    \E local, remote \in Peers :
+        \/ ACTIVE_OPEN(local, remote)
+        \/ PASSIVE_OPEN(local, remote)
+        \/ CLOSE_SYN_SENT(local, remote)
+        \/ CLOSE_SYN_RECEIVED(local, remote)
+        \/ CLOSE_LISTEN(local, remote)
+        \/ CLOSE_ESTABLISHED(local, remote)
+        \/ CLOSE_CLOSE_WAIT(local, remote)
+        \/ SEND(local, remote)
+
+-----------------------------------------------------------------------------
+       
+SynSent(local, remote) ==
+    /\ local # remote
+    /\ connstate[local] = "SYN-SENT"
+    /\ UNCHANGED tcb
+    /\ \/ /\ IsPrefix(<<"SYN">>, network[local])
```

Comment:
Prefer Head(network[local]) = "SYN" over IsPrefix?

Reply:
I find `network[local] # <<>> /\ Head(network[local]) = "SYN"` rather verbose. Moreover, I wanted the actions to be consistent, and `Note2` definitely requires `IsPrefix`.  Would be interesting to specify TCP in Distributed PCal. :-)